### PR TITLE
Generate reasonable markup even if tags have unexpected case.

### LINKED ISSUE #2756: Warn if rendering an HTML tag spelled with uppercase letters
If I try running the following code:

```javascript
var Hello = React.createClass({
    render: function() {
      return React.createElement('BR', null);
    }
});
 
React.render(<Hello/>, document.body);
```

I get two `<br>` tags inserted into the dom, instead of one. I made a quick jsfiddle to show this. http://jsfiddle.net/kb3gN/8662/

I'm running Chrome, if that's at all relevant.

### PR DESCRIPTION
**Pull request by jisaacks**
Fixes #2756

### CODE REVIEW

**Comment by jimfb** [comment]
lgtm.  cc @spicyj.

**Comment by spicyj** [comment]
1. This should use the `warning` module, not a call to `console.error`.

2. This shouldn't warn more than once per tag. Probably easiest to combine with validateDangerousTag (rename to just validateTagName) and take advantage of validatedTagCache.

3. SVG tags like `fontFace` start with a lowercase letter and are valid. Let's only warn when the first letter is uppercase. Let's use the message: `"DOM tag names should be lowercase but got: %s"`. 

When you're done, can you squash your commits?

**Comment by jisaacks** [comment]
@spicyj I think that covers everything. Let me know if there is anything else :)

**Comment by sebmarkbage** [comment]
I'm not sure this is the right solution. The HTML DOM allows case insensitive definitions and it might even be encouraged to use capitalization for certain custom elements.

Shouldn't we just make our `omittedCloseTags` check case-insensitive?

**Comment by jisaacks** [comment]
@sebmarkbage while that would fix the actual bug, I think it would be confused to users, that `DIV`, `UL`, `A

**Comment by jisaacks** [comment]
@sebmarkbage while that would fix the actual bug, I think it would be confusing to users, that `DIV`, `UL`, `A` etc.. are all valid but `BR`, `IMG`, etc.. are not.

What would the error message say?  __Self closing elements must be lowercase!__

> ...it might even be encouraged to use capitalization for certain custom elements.

Are custom elements even supported by React? I thought it was just these [dom nodes](https://github.com/facebook/react/blob/master/src/renderers/dom/shared/ReactDOMComponent.js#L107)

If so, maybe allow the different cases but downcase the tag name when checking it [here](https://github.com/facebook/react/blob/master/src/renderers/dom/shared/ReactDOMComponent.js#L533) and [here](https://github.com/facebook/react/blob/master/src/renderers/dom/shared/ReactDOMComponent.js#L107). Or only check case sensitivity on the dom nodes defined in the linked above.

Either way, enforcing strict casing on a **subset** of the DOM, because of a bug feels wrong.

**Comment by sebmarkbage** [comment]
I didn't mean that we should enforce a strict casing on a subset. We should not enforce it on anything.

I.e. downcase here https://github.com/facebook/react/blob/master/src/renderers/dom/shared/ReactDOMComponent.js#L533

**Comment by jisaacks** [comment]
@sebmarkbage got it. Should I make that change now, or does @spicyj need to sign off? Just not sure how the process here works, and don't want to waste effort. Thanks!

**Comment by jimfb** [comment]
@sebmarkbage is the ultimate authority, so a go-ahead from him is generally enough to do anything :).

I'm fairly certain that @spicyj won't object to this proposed solution.

Go ahead and make the change :).

**Comment by spicyj** [comment]
No objections here.

**Comment by spicyj** [comment]
@sebmarkbage doesn't always win!

**Comment by jimfb** [comment]
Haha, true, but I'm trying to remember the last time he was overruled.  It doesn't happen very often.

**Comment by jimfb** [comment]
It's good though, because he is usually right, and is always willing to hear everyone's opinion/feedback before committing to a course of action.

**Comment by sebmarkbage** [comment]
$40 says I don't always win.

**Comment by jimfb** [comment]
That's true.  Note to self: never bet money against @spicyj.

**Comment by sebmarkbage** [comment]
I'm always right because I change my mind. Doesn't mean you should listen to my initial inclination. Unlike American politics I think we should value new input. :)

**Comment by jimfb** [comment]
ok, ok, enough fun.  New diff is up, back on topic.

@jisaacks
I think you missed this one:
https://github.com/facebook/react/blob/master/src/renderers/dom/shared/ReactDOMComponent.js#L224

**Comment by spicyj** [comment]
All the switch statements too for custom form components (incl. `this._tag === 'select'`).

**Comment by jisaacks** [comment]
@jimfb I will fix that right now. Already fixed the trailing comma, which confuses me.

@spicyj um.. what?

**Comment by jimfb** [comment]
@jisaacks: What @spicyj was trying to say is that there are other cases in the codebase where we check the value of the tag.  Since now we're supporting arbitrary case, we need to fix those also.

**Comment by spicyj** [comment]
And a bunch that look like this: https://github.com/facebook/react/blob/c5fb3ff9870cc09a6ec82672e854ab54a412cef1/src/renderers/dom/shared/ReactDOMComponent.js#L677

**Comment by jisaacks** [comment]
OK new commit coming in just a minute.

**Comment by jisaacks** [comment]
OK @jimfb @spicyj @sebmarkbage I think that covers it.

**Comment by sebmarkbage** [comment]
How about making it simple for yourself and just do it once `this._tag = tag.toLowerCase();`?

This is already going to have perf implications so we should minimize this work by just doing it once.

Optionally we could use the original tag here to make sure case is preserved in the DOM: https://github.com/facebook/react/blob/c5fb3ff9870cc09a6ec82672e854ab54a412cef1/src/renderers/dom/shared/ReactDOMComponent.js#L553

**Comment by spicyj** [comment]
We at least need to preserve case for SVG tags.

**Comment by sebmarkbage** [comment]
@spicyj Which browser requires that? They should be insensitive according to spec.

Or are you referring to #4077 for XHTML in which case we need it to be sensitive for XHTML as well?

**Comment by jisaacks** [comment]
> Optionally we could use the original tag here to make sure case is preserved in the DOM

So then we would need to store something like:

```javascript
this._tag = tag.toLowerCase();
this._originalTag = tag;
```

In order to use it later. Unless I am missing something.

**Comment by sebmarkbage** [comment]
Yea, that works. It is just a bit unnecessary to store that piece of state for every node but we can refactor that later. You could also use `this._currentElement.type` to get the original I guess.

**Comment by spicyj** [comment]
> They should be insensitive according to spec.

TIL. I guess we might want to support XHTML though. Sigh. Browsers.

**Comment by sebmarkbage** [comment]
Is XHTML really a thing yet? Are browsers even supporting it? Maybe we should have a special build for that. If we support XHTML then they aren't insensitive anymore.

**Comment by jisaacks** [comment]
@sebmarkbage let me know if that looks good.

**Comment by jimfb** [comment]
To keep things maximally general, it seems intuitively better to avoid mangling the input whenever possible.  The more transparent we can be (ie. the closer we can be to passing inputs straight through) the better.  I could see people wanting case sensitivity, as we see with things like SVG and XHTML and WebComponents.

That said, I'm not strongly attached to preserving case sensitivity.  This one doesn't bother me quite as much as the unitless numbers bug does.

In this latest diff, it looks like we're using upper case for the open tag and lower case for the closing tag?  That can't be right :P, unless I'm miss-reading something?

**Comment by sebmarkbage** [comment]
Yea, we should probably use the same for the closing tag too. I figured it didn't matter because of case-insensitive parsing and we only really want to preserve the case for purposes of debugging since it shouldn't matter for anything else.

However, for safety, and our own well being we should probably use the same case in the closing case too.

File: src/renderers/dom/shared/ReactDOMComponent.js [hunk 1]
```diff
@@ -550,7 +550,7 @@ ReactDOMComponent.Mixin = {
    * @return {string} Markup of opening tag.
    */
   _createOpenTagMarkupAndPutListeners: function(transaction, props) {
-    var ret = '<' + this._tag;
+    var ret = '<' + this._currentElement.type;
```
Comment (sebmarkbage) [inline src/renderers/dom/shared/ReactDOMComponent.js]: We need this to line up with https://github.com/facebook/react/blob/master/src/renderers/dom/shared/ReactDOMComponent.js#L536

Comment (jisaacks) [inline src/renderers/dom/shared/ReactDOMComponent.js]: Yep Alright pushed that change up :)

**Comment by jisaacks** [comment]
@sebmarkbage @jimfb closing tag should match now :)

**Comment by jimfb** [comment]
Ok, thanks @jisaacks!  Sorry about all the back and forth, but this is a good change!  Much appreciated!

**Comment by jisaacks** [comment]
@jimfb no worries! Thanks! :)

**Pull request merged**